src/common/FastRational.h

```diff
@@ -572,52 +599,75 @@ inline bool FastRational::isWellFormed() const
 {
     return (  wordPartValid() || mpqPartValid() )
            && ( !wordPartValid() || (den != 0 && gcd(absVal(num), den)==1) )
-           && ( !mpqPartValid()  || mpz_sgn(mpq_denref(mpq))!=0 );
+           && ( !mpqPartValid()  || mpz_sgn(mpq_denref(mpq))!=0 )
+           && ( !(wordPartValid() && mpqPartValid()) || wordAndMpqEqual());
+    // Check that if both wordPartValid() and mpqPartValid(), the are the same number
+}
+
+inline bool FastRational::wordAndMpqEqual() const {
+    assert(wordPartValid() && mpqPartValid());
+    uword int_den = mpz_get_ui(mpq_denref(mpq));
+    word int_num = mpz_get_si(mpq_numref(mpq));
+    return int_den == den and int_num == num;
 }
 
 inline FastRational::FastRational(word n, uword d) : state{State::WORD_VALID} {
     assert(d > 0);
     if (n == 0) {
         num = 0;
         den = 1;
-    } else if (n > 0) {
-        word common = gcd(uword(n), d);
-        num = n/common;
-        den = d/common;
     } else {
-        word common = gcd(uword(-n), d);
-        num = n/common;
-        den = d/common;
+        uword common = gcd<uword>(absVal(n), d);
+        if (common > 1) {
+            num = n / common;
+            den = d / common;
+        } else {
+            num = n;
+            den = d;
+        }
     }
 }
 
 inline void addition(FastRational& dst, const FastRational& a, const FastRational& b) {
     if (a.wordPartValid() && b.wordPartValid()) {
         if (b.num == 0) {
-            dst.num = a.num;
-            dst.den = a.den;
+            dst = a;
```

Comment:
I wanted to write I would keep the original, but then you get into questions like, what happens if `dst` had the `mpq` part allocated...

Reply:
Changed back to original